kimchi/src/snarky/cvar.rs

```diff
@@ -0,0 +1,329 @@
+                b.eval_inner(context, scale, res);
+            }
+            CVar::Scale(s, v) => {
+                v.eval_inner(context, scale * s, res);
+            }
+        }
+    }
+
+    /// Evaluate the field element associated to a variable (used during witness generation)
+    pub fn eval(&self, context: &impl (Fn(usize) -> F)) -> F {
+        let mut res = F::zero();
+        self.eval_inner(context, F::one(), &mut res);
+        res
+    }
+
+    pub fn to_constant_and_terms_inner(
+        &self,
+        scale: F,
+        constant: F,
+        terms: Vec<Term<F>>,
+    ) -> (F, Vec<Term<F>>) {
+        match self {
+            CVar::Constant(c) => (constant + (scale * c), terms),
+            CVar::Var(v) => {
+                let mut new_terms = vec![(scale, *v)];
+                new_terms.extend(terms);
+                (constant, new_terms)
+            }
+            CVar::Scale(s, t) => t.to_constant_and_terms_inner(scale * s, constant, terms),
+            CVar::Add(x1, x2) => {
+                let (c1, terms1) = x1.to_constant_and_terms_inner(scale, constant, terms);
+                x2.to_constant_and_terms_inner(scale, c1, terms1)
+            }
+        }
+    }
+
+    pub fn to_constant_and_terms(&self) -> (Option<F>, Vec<Term<F>>) {
+        let (constant, terms) = self.to_constant_and_terms_inner(F::one(), F::zero(), vec![]);
+        let constant = if constant.is_zero() {
+            None
+        } else {
+            Some(constant)
+        };
+        (constant, terms)
+    }
+
+    pub fn scale(&self, scalar: F) -> Self {
+        if scalar.is_zero() {
+            return CVar::Constant(scalar);
+        } else if scalar.is_one() {
+            return self.clone();
+        }
+
+        match self {
+            CVar::Constant(x) => CVar::Constant(*x * scalar),
+            CVar::Scale(s, v) => CVar::Scale(*s * scalar, v.clone()),
+            CVar::Var(_) | CVar::Add(..) => CVar::Scale(scalar, Box::new(self.clone())),
+        }
+    }
+
+    pub fn linear_combination(terms: &[ScaledCVar<F>]) -> Self {
+        let mut res = CVar::Constant(F::zero());
+        for (cst, term) in terms {
+            res = res.add(&term.scale(*cst));
+        }
+        res
+    }
+
+    pub fn sum(vs: &[&Self]) -> Self {
+        let terms: Vec<_> = vs.iter().map(|v| (F::one(), (*v).clone())).collect();
+        Self::linear_combination(&terms)
+    }
+
+    pub fn mul(&self, other: &Self, label: Option<&'static str>, cs: &mut RunState<F>) -> Self {
+        match (self, other) {
+            (CVar::Constant(x), CVar::Constant(y)) => CVar::Constant(*x * y),
+
+            // TODO: this was not in the original ocaml code, but seems correct to me
+            (CVar::Constant(cst), _) | (_, CVar::Constant(cst)) if cst.is_zero() => {
+                CVar::Constant(F::zero())
+            }
+
+            // TODO: same here
+            (CVar::Constant(cst), cvar) | (cvar, CVar::Constant(cst)) if cst.is_one() => {
```

Comment:
Delete.

Reply:
this logic was actually in the `scale` function!